### DISCUSSION TITLE
STB_USE_SYSTEM_NODE gives confusing errors

### DISCUSSION THREAD

**Discussion by mgorny** [pradyunsg/sphinx-theme-builder#31] [Potential Issue] [answered]
I am attempting to build `furo` with `STB_USE_SYSTEM_NODE=true`, and I'm getting the following message:

```
$ python -m build -w
* Creating virtualenv isolated environment...
* Installing packages in isolated environment... (sphinx-theme-builder >= 0.2.0a10)
* Getting dependencies for wheel...
* Building wheel...
[stb] # nodeenv does not exist.
[stb] # Generating new nodeenv with NodeJS 16.15.1!
[stb] # Will use system nodeJS, since version matches.
[stb] $ python -m nodeenv --node=system --prebuilt --clean-src /tmp/furo/.nodeenv
/tmp/build-env-a3fsfm3s/lib/python3.11/site-packages/nodeenv.py:26: DeprecationWarning: 'pipes' is deprecated and slated for removal in Python 3.13
  import pipes
which: no nodejs in (/tmp/build-env-a3fsfm3s/bin:/home/mgorny/perl5/bin:/home/mgorny/node_modules/.bin/:/home/mgorny/bin:/usr/i686-pc-linux-gnu/gcc-bin/4.9.2:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/bin:/usr/lib/llvm/14/bin:/usr/lib/llvm/9/bin)
[stb] (nodeenv) $ node --version
v18.4.0
error: nodeenv-version-mismatch

× The `nodeenv` for this project is unhealthy.
╰─> There is a mismatch between what is present in the environment (v18.4.0) and the expected version of NodeJS (v16.15.1).

hint: Deleting the .nodeenv directory and trying again may work.

Link: https://sphinx-theme-builder.rtfd.io/errors/#nodeenv-version-mismatch

ERROR Backend subprocess exited when trying to invoke build_wheel
```

This is entirely confusing — first it says that the version matches, then it says it is mismatched. It tells me to remove `.nodeenv` but that doesn't change anything. Finally, it links to a doc that's 404.

By the way, the 'source' link on pypi is also 404.

**Comment by pradyunsg** [comment]
Both the broken links are fixed on `main`, IIRC.

https://sphinx-theme-builder.readthedocs.io/en/latest/errors/#nodeenv-version-mismatch

Yea, the message shouldn't say ", since version matches" since that isn't being checked in the logic prior anymore.

**Reply by mgorny** [reply]
So what is the correct solution here?

**Reply by pradyunsg** [reply]
Match the system-provided NodeJS with the NodeJS version as defined by the theme. For example, you're using a different major, minor and patch version of NodeJS to build Furo in your example.

**Reply by mgorny** [reply]
So, practically saying, the only way to use system NodeJS is to have every single NodeJS version ever released packaged and replace installed version (presumably breaking the rest of the system) every single time I'm about to build a different theme using sphinx-theme-builder. Is that correct?

**Reply by pradyunsg** [reply]
If you continue to ask rhetorical slightly-accusatory questions, this conversation is going to end quickly. :)

I can understand that enforcing a complete version match might be a bit much, but you're using a different major version of NodeJS for the build. I don't think we'd ever want to allow that. At best, [your build going to pass with weird unknown auxilary implications](https://blog.yossarian.net/2021/02/28/Weird-architectures-werent-supported-to-begin-with) or it's going to fail. This check serves to prevent redistributors from building packages in a way that upstream maintainers cannot sustainably support.

**Reply by mgorny** [reply]
I'm afraid this design has major security implications. I mean, if a package pins to a specific furo version, and furo effectively pins to a specific nodejs version, then the specific nodejs version turns out to have major vulnerabilities, anyone installing this version from source is susceptible to running vulnerable software on their system.

Would including generated assets in sdist, so that rebuilding them could be skipped be an option?

**Reply by arthurzam** [reply]
Would it be possible to lessen the exact version dependency on nodejs? If you want exact major version - fine. But also exact minor, etc is too narrow. This affects badly distribution who use source build.

I have carefully read your linked article (thank you BTW, it was an interesting read), but I'm not sure how does it concern us in this discussion? We aren't talking here about "niche" architecture types, nor about using another "compiler" family. We are concerned by the exact pinned version of nodejs, which might be vulnerable, bugged, etc.

I think it would be a good compromise if we could have a selected exact major version and a minimal minor.

**Reply by pradyunsg** [reply]
> Would it be possible to lessen the exact version dependency on nodejs? 

Yup, happy to.

> I think it would be a good compromise if we could have a selected exact major version and a minimal minor.

Seems reasonable to me!

**Reply by pradyunsg** [reply]
> Would including generated assets in sdist, so that rebuilding them could be skipped be an option?

No. The assets are explicitly generated during wheel generation rather than the source distribution, to minimise the differences between what the sdist contains vs what's under version control. This is also necessary to enable pip installing directly from the git repository.

**Reply by pradyunsg** [reply]
> We aren't talking here about "niche" architecture types, nor about using another "compiler" family. We are concerned by the exact pinned version of nodejs, which might be vulnerable, bugged, etc.

Well, the broader point that the article is making is that "it builds and tests pass" != "it works correctly".

FWIW, a large part of my motivation to start strict here is that diagnosing issues with Sphinx themes is a non-trivial problem on its own, requiring the author to spend time and effort in multiple knowledge domains, across with multiple languages and a matrix of libraries across them. Anything that I can do to restrict the combinational explosion on this front and making them explicit failures, makes life easier for the theme authors and their users.

Disallowing this mismatch avoids the potential for a failure modes that look like "Gentoo/Debian/some-other-linux-distro built the theme with NodeJS 16 instead of NodeJS 18, resulting in a certain dependency used in the JS build being a different version, which results in differences in the stylesheets / javascript being generated, resulting in an esoteric failure that the upstream author cannot reproduce". In this hypothetical, the upstream maintainers really have no way of diagnosing the underlying issue (without a significant investment of time) and there's basically nothing they can do beyond maybe providing the user with a workaround or asking them to not use the broken copy from the redistributor. That's assuming the redistributor isn't meddling with the NodeJS dependencies to get the theme's build to work, because it failed on the first attempt (again: it builds != it works correctly) which would mean that it's even harder to pinpoint the issue for upstream.

The best case scenario is that the redistributor realises what they're doing is bound to be broken, and asks for NodeJS 16 support from upstream. And, that's what they'll have to do with this being an explicit failure (I hope?) -- they just don't have the option to significantly raise maintainance costs for the upstream developers, just because they're able to massage a build system into building without failures.

**Reply by arthurzam** [reply]
> > We aren't talking here about "niche" architecture types, nor about using another "compiler" family. We are concerned by the exact pinned version of nodejs, which might be vulnerable, bugged, etc.
> 
> Well, the broader point that the article is making is that "it builds and tests pass" != "it works correctly".
> 
> FWIW, a large part of my motivation to start strict here is that diagnosing issues with Sphinx themes is a non-trivial problem on its own, requiring the author to spend time and effort in multiple knowledge domains, across with multiple languages and a matrix of libraries across them. Anything that I can do to restrict the combinational explosion on this front and making them explicit failures, makes life easier for the theme authors and their users.
> 
> Disallowing this mismatch avoids the potential for a failure modes that look like "Gentoo/Debian/some-other-linux-distro built the theme with NodeJS 16 instead of NodeJS 18, resulting in a certain dependency used in the JS build being a different version, which results in differences in the stylesheets / javascript being generated, resulting in an esoteric failure that the upstream author cannot reproduce". In this hypothetical, the upstream maintainers really have no way of diagnosing the underlying issue (without a significant investment of time) and there's basically nothing they can do beyond maybe providing the user with a workaround or asking them to not use the broken copy from the redistributor. That's assuming the redistributor isn't meddling with the NodeJS dependencies to get the theme's build to work, because it failed on the first attempt (again: it builds != it works correctly) which would mean that it's even harder to pinpoint the issue for upstream.
> 
> The best case scenario is that the redistributor realises what they're doing is bound to be broken, and asks for NodeJS 16 support from upstream. And, that's what they'll have to do with this being an explicit failure (I hope?) -- they just don't have the option to significantly raise maintainance costs for the upstream developers, just because they're able to massage a build system into building without failures.

Thank you for the thorough reply to my questions!

**Comment by pradyunsg** [comment]
> Would including generated assets in sdist, so that rebuilding them could be skipped be an option?

@mgorny If you're fine with using generated assets from the JS build, would it be fine if each theme provided a zip file which you can unpack to install the theme (with the build assets and all)?

If so, that's *exactly* what the wheel generated for each theme is. Any Python package that ends with a `-none-any.whl` should be a version-agnostic package that's basically installed by unpacking the zip file and generating a wrapper.

**Comment by pradyunsg** [comment, marked as answer]
I've filed #32 as a follow up on the discussion here. Thanks for flagging this concern @mgorny! ^>^
